--- git status ---
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env.py
	modified:   source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env_cfg.py

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	"pr\303\263by in\305\274ynierka nagrania zmiany/pr\303\263ba komendy 0.mp4"

no changes added to commit (use "git add" and/or "git commit -a") 


--- git diff ---
diff --git a/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env.py b/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env.py
index c014cd1..efdb99a 100644
--- a/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env.py
+++ b/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env.py
@@ -243,6 +243,12 @@ class InzynierkaizaklaboratoriumEnv(DirectRLEnv):
 
                 "stand_lin_vel",
                 "stand_yaw_vel",
+                "stand_upright",
+                "stand_tilt_debug",
+                "stand_action",
+                "stand_double_contact_debug",
+                "stand_single_support_debug",
+                "stand_single_support",
             ]
         }
 
@@ -568,8 +574,21 @@ class InzynierkaizaklaboratoriumEnv(DirectRLEnv):
         # Aktywne wyłącznie dla zerowej komendy.
         # ============================================================
 
+        stand_pose_excess = torch.clamp(
+            torch.abs(joint_delta_all) - 0.10,
+            min=0.0,
+        )
+
         stand_pose_error = torch.mean(
-            torch.square(joint_delta_all),
+            torch.square(stand_pose_excess),
+            dim=1,
+        )
+        stand_tilt_error = torch.sum(
+            torch.square(self._robot.data.projected_gravity_b[:, :2]),
+            dim=1,
+        )
+        stand_action_error = torch.mean(
+            torch.square(self._actions),
             dim=1,
         )
 
@@ -997,6 +1016,10 @@ class InzynierkaizaklaboratoriumEnv(DirectRLEnv):
         double_support = left_contact & right_contact
 
         stand_double_contact = double_support.float()
+
+        stand_single_support = (
+                left_contact ^ right_contact
+        ).float()
         # ============================================================
         # PRZEDŁUŻONY DOUBLE SUPPORT
         #
@@ -1754,6 +1777,24 @@ class InzynierkaizaklaboratoriumEnv(DirectRLEnv):
                     * stand_gate
                     * self.step_dt
             ),
+            "stand_upright": (
+                    self.cfg.rew_scale_stand_upright
+                    * stand_tilt_error
+                    * stand_gate
+                    * self.step_dt
+            ),
+            "stand_action": (
+                    self.cfg.rew_scale_stand_action
+                    * stand_action_error
+                    * stand_gate
+                    * self.step_dt
+            ),
+            "stand_single_support": (
+                    self.cfg.rew_scale_stand_single_support
+                    * stand_single_support
+                    * stand_gate
+                    * self.step_dt
+            ),
 
         }
 
@@ -1930,6 +1971,22 @@ class InzynierkaizaklaboratoriumEnv(DirectRLEnv):
                 * stand_gate
                 * self.step_dt
         )
+        self._episode_sums["stand_tilt_debug"] += (
+                stand_tilt_error
+                * stand_gate
+                * self.step_dt
+        )
+        self._episode_sums["stand_double_contact_debug"] += (
+                stand_double_contact
+                * stand_gate
+                * self.step_dt
+        )
+
+        self._episode_sums["stand_single_support_debug"] += (
+                stand_single_support
+                * stand_gate
+                * self.step_dt
+        )
         return reward
 
     def _get_dones(self) -> tuple[torch.Tensor, torch.Tensor]:
diff --git a/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env_cfg.py b/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env_cfg.py
index 47e2176..00f7517 100644
--- a/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env_cfg.py
+++ b/source/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium/tasks/direct/inzynierkaizaklaboratorium/inzynierkaizaklaboratorium_env_cfg.py
@@ -129,7 +129,7 @@ class InzynierkaizaklaboratoriumEnvCfg(DirectRLEnvCfg):
 
     # 15% wylosowanych komend = stanie w miejscu.
     standing_command_probability = 0.25
-    command_resample_time_s = 2.5
+    command_resample_time_s = 10
 
     # Gotowe zakresy na późniejszy etap omnidirectional.
     omni_min_speed = 0.08
@@ -218,10 +218,14 @@ class InzynierkaizaklaboratoriumEnvCfg(DirectRLEnvCfg):
     stand_pose_sigma = 0.025
     stand_height_sigma = 0.025
 
-    rew_scale_stand_pose = -6.0
-    rew_scale_stand_height = 2.0
-    rew_scale_stand_joint_vel = -0.5
-    rew_scale_stand_double_contact = 1.0
-
-    rew_scale_stand_lin_vel = -8.0
-    rew_scale_stand_yaw_vel = -3.0
+    rew_scale_stand_pose = 0.0
+    rew_scale_stand_height = 0.0
+    rew_scale_stand_joint_vel = -0.05
+    rew_scale_stand_double_contact = 2.0
+
+    rew_scale_stand_lin_vel = -10.0
+    rew_scale_stand_yaw_vel = -4.0
+    rew_scale_stand_upright = -10.0
+    rew_scale_stand_action = -1.0
+    # CHCEMY DWIE STOPY NA ZIEMI.
+    rew_scale_stand_single_support = -2.0